### DISCUSSION TITLE
Clone the repository and use enviroment variables

### DISCUSSION THREAD

**Discussion by fabiantorress** [twilio-professional-services/flex-project-template#506] [Q&A] [answered]
Hi, someone has experience using the plugin-flex-ts-template-v2 repo with enviromental variables? I am newbie using Typescript and I try to use .env files in production but I am getting undefined. I just cloned the repo and ereased the features that I wasn't interested.
Specifically I want to use the chat-transfer feature and the process.env.FLEX_APP_SERVERLESS_FUNCTONS_DOMAIN variable is getting undefined. The variable is defined in my .env file.
Thanks in advanced.

**Comment by dremin** [comment]
Hi @fabiantorress! If you only need chat transfer, have you tried using the plugin library version instead (accessed via Flex UI > Admin > Plugins > Plugin library tab)? The plugin library version was derived from the template version, so it should be more-or-less the same.

If you'd like to go forward with using the template instead: You actually should not need to manually modify the .env files, as the template sets these automatically based on the items existing in your Twilio account. I would suggest following these steps to ensure success:
1. Remove all features except the one you want [using these instructions](https://twilio-professional-services.github.io/flex-project-template/getting-started/use-specific-feature). Then, push those changes to your repo.
2. Deploy the template [using the instructions here](https://twilio-professional-services.github.io/flex-project-template/getting-started/install-template). Note specifically on step 6, be sure to select the options that this is an initial deploy, and to deploy terraform. This ensures the chat transfer workflow gets created and automatically populated into the .env during deployment.
3. Now that you have deployed the template, we can make it work locally as well. If you have not tried running locally yet, [follow these directions](https://twilio-professional-services.github.io/flex-project-template/getting-started/run-locally). Otherwise, if you have already installed locally, you can run `npm run postinstall -- --overwrite` and the template scripting will update your .env files with the appropriate values.

If you have more questions about how the .env files work and how configuration is managed in the template in general, [this docs page](https://twilio-professional-services.github.io/flex-project-template/building/template-utilities/configuration) describes how it all works.

I hope this helps!

**Reply by fabiantorress** [reply]
What about the FLEX_APP_SERVERLESS_FUNCTONS_DOMAIN variable? Is there a way to set this in the codebase without deploying? Locally the chat-transfer works because it used the appConfig.js configuration but when I deploy to Twilio it must use the .env file.
I am trying to avoid Terraform deploy because this will set up TaskRouter and Studio configuration and I do not want to break something that is running in production. 
By the way, I am using the Legacy Programmable Chat, that is why I am not using the official plugin.

**Reply by dremin** [reply]
@fabiantorress

For the serverless domain variable, this cannot be populated until the serverless service has been deployed at least once, since the domain name is dynamically generated to be unique to your account.

You can skip Terraform, that is understandable. If you skip Terraform, you will need to manually create the chat transfer TaskRouter workflow [using this example workflow JSON as an example](https://github.com/twilio-professional-services/flex-project-template/blob/main/plugin-flex-ts-template-v2/src/feature-library/chat-transfer/example-taskrouter-workflow.json), using either API or the Twilio Console. The deploy scripting will pick it up automatically as long as the workflow is named `Chat Transfer`. Once you have created this workflow, you may then run the GitHub Actions deploy with the terraform option disabled.

**Reply by fabiantorress** [reply]
Yes, my serverless service is already deployed (I have the url). I have already my Workflow as the example. The only issue that I am having is the FLEX_APP_SERVERLESS_FUNCTONS_DOMAIN variable that is not reading my .env file. I want to use the enviromental variable because I have different stages (prod, dev) and I just want to change the serverless url in one place (the .env file).

**Reply by dremin** [reply]
@fabiantorress In that case, you should be able to run the GitHub Actions deploy script, with terraform disabled, and the .env will automatically populate correctly for the selected environment.

**Reply by fabiantorress** [reply]
@dremin I understand, but to be clear, there is no way I can use .env files and based on my stage project link the corresponding env file with stage? Like this: https://www.twilio.com/es-mx/docs/flex/developer/plugins/environment-variables . I already have done this with my custom plugins but using Javascript, not Typescript.

**Reply by dremin** [reply]
@fabiantorress You can use .env files like any other plugin, however, we recommend using other facilities for configuration, as .env files require a re-deployment to modify.

**Reply by fabiantorress** [reply, marked as answer]
I didn't fix the undefined serverless_functions_domain that I was getting, but I got it to work using the Flex UI Configuration as the documentation suggested. Thanks @dremin !
